evm/src/keccak_sponge/columns.rs

```diff
@@ -0,0 +1,118 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::mem::{size_of, transmute};
+
+use crate::util::{indices_arr, transmute_no_compile_time_size_checks};
+
+pub(crate) const KECCAK_WIDTH_BYTES: usize = 200;
+pub(crate) const KECCAK_WIDTH_U32S: usize = KECCAK_WIDTH_BYTES / 4;
+pub(crate) const KECCAK_RATE_BITS: usize = 1088;
+pub(crate) const KECCAK_RATE_BYTES: usize = KECCAK_RATE_BITS / 8;
+pub(crate) const KECCAK_RATE_U32S: usize = KECCAK_RATE_BYTES / 4;
+pub(crate) const KECCAK_CAPACITY_BYTES: usize = 64;
+pub(crate) const KECCAK_CAPACITY_U32S: usize = KECCAK_CAPACITY_BYTES / 4;
+
+#[repr(C)]
+#[derive(Eq, PartialEq, Debug)]
+pub(crate) struct KeccakSpongeColumnsView<T: Copy> {
+    /// 1 if this row represents a dummy operation (for padding the table); 0 otherwise.
+    pub is_dummy: T,
+
+    /// 1 if this row represents a full input block, i.e. one in which each byte is an input byte,
+    /// not a padding byte; 0 otherwise.
+    pub is_full_input_block: T,
+
+    /// 1 if this row represents the final block of a sponge, in which case some or all of the bytes
+    /// in the block will be padding bytes; 0 otherwise.
+    pub is_final_block: T,
+
+    // The address at which we will read the input block.
+    pub context: T,
+    pub segment: T,
+    pub virt: T,
+
+    /// The timestamp at which inputs should be read from memory.
+    pub timestamp: T,
+
+    /// The length of the original input, in bytes.
+    pub len: T,
+
+    /// The number of input bytes that have already been absorbed prior to this block.
+    pub already_absorbed_bytes: T,
+
+    /// If this row represents a final block row, the `i`th entry should be 1 if the final chunk of
+    /// input has length `i` (in other words if `len - already_absorbed == i`), otherwise 0.
+    ///
+    /// If this row represents a full input block, this should contain all 0s.
+    pub is_final_input_len: [T; KECCAK_RATE_BYTES],
+
+    /// The initial rate bits of the sponge, at the start of this step.
+    pub original_rate_bits: [T; KECCAK_RATE_BITS],
```

Comment:
Could we use a lookup table to compute the XORs and have these be 4/8-bit chunks instead?

Reply:
Hm by inner table do you mean prover-provided (instead of preprocessed) table?

I think lookup tables could be better, but perhaps not by a significant margin, depending on our table size. Like with 2 cols for input chunks, 1 for the output chunk, 2 permuted cols, and 3 `Z` cols, we'd be at 8 cols, giving similar costs as having 8 input bit cols.

I like the idea of using the logic table, only slight complication is that it uses 256-bit inputs, which doesn't divide the Keccak rate (1088 bits). We could reduce its inputs to 64 bits, though that'd affect the CPU's CTL costs a bit. Maybe best to keep it at 256 bits and just do 5 lookups here (the last would only use 64 bits; we could optimize it later). Does that sound good?

Reply:
> Hm by inner table do you mean prover-provided (instead of preprocessed) table?

Yes, similar to how we do range checks.

> I think lookup tables could be better, but perhaps not by a significant margin, depending on our table size. Like with 2 cols for input chunks, 1 for the output chunk, 2 permuted cols, and 3 `Z` cols, we'd be at 8 cols, giving similar costs as having 8 input bit cols.

Yeah I agree for 4-bit chunks it might not be worth it. For 8-bit chunks it might already be better, but then we'd need to pad the table to have a length of 2^16, which might be wasteful.
 
> I like the idea of using the logic table, only slight complication is that it uses 256-bit inputs, which doesn't divide the Keccak rate (1088 bits). We could reduce its inputs to 64 bits, though that'd affect the CPU's CTL costs a bit. Maybe best to keep it at 256 bits and just do 5 lookups here (the last would only use 64 bits; we could optimize it later). Does that sound good?

Yes sounds good! We can store the rate as `u32`s and then use a `Column::le_u32s` in the 5 CTLs to the logic table.

Reply:
Ok I made the change. I stored the previous rate part as `u32`s and the block as `u8`s (since they're `u8`s in memory). The logic CTL uses `u32`s, so there's no conversion needed for the previous rate part, but I used `le_bytes` to combine block bytes.